reference/reference.go

```diff
@@ -56,6 +65,11 @@ var (
 type Reference interface {
 	// String returns the full reference
 	String() string
+
+	// internal requires that all instances of Reference must be implemented
+	// by this package. This allows us to ensure that instances can be treated
+	// as an algebraic data type.
+	internal()
```

Comment:
I am sorry, but I want to force that requirement. I am not to thrilled about the forking of the reference package and have put in measures to ensure that doesn't happen again. This also let's us write helper functions with stronger guarantees.